[agent] Support for running secrets backends with sha256 verification

xornivore opened the pull request:
### What does this PR do?

- Adds support for `secret_backend_command_sha256` that enforces SHA256 of the secret backend binary.

Note that when `secret_backend_command_sha256` is used the `datadog.yaml` config has to be restricted as follows:
- Must be owned by root and cannot have writable others' permissions on *nix.
- Cannot be writable by principals other than `Local System`, `Administrators` or `ddagentuser` on Windows.

### Motivation

- Provide cross platform way to enforce the validity of the `secret_backend_command` binary.

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

- Validate existing secrets backend functionality still works as expected.
- Validate configuration of `secret_backend_command_sha256` works correctly on Linux and Windows.

In order to validate this functionality on Windows the datadog.yaml file has to be configured to disallow writes by principals other than `Local System`, `Administrators`, or `ddagentuser`.

The secret backend binary should have access by the account running the agent.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

xornivore committed:
[agent] Support for running secrets backends with sha256 verification

- Adds support for `secret_backend_command_sha256` that enforces SHA256 of the secret backend binary.
- Allows invoking secret backends in process agent on Windows as long as the SHA256 is provided, and the config file has a restricted ACL.
- Address review comments - nolint + fix windows ACE check

clarkb7 commented on pkg/secrets/fetch_secret.go:
```diff
@@ -49,6 +54,35 @@ func execCommand(inputPayload string) ([]byte, error) {
+		if err != nil {
+			return nil, err
+		}
+		defer func() { _ = f.Close() }()
```
you can use `//nolint` instead of wrapping the call
```
defer f.Close() //nolint:errcheck
```

xornivore committed:
Resolve ddagentuser when running as Local System so the DACL check can work for both

just-chillin commented on pkg/config/config.go:
```diff
@@ -1601,6 +1602,8 @@ func ResolveSecrets(config Config, origin string) error {
 		config.GetInt("secret_backend_output_max_size"),
 		config.GetBool("secret_backend_command_allow_group_exec_perm"),
+		config.GetString("secret_backend_command_sha256"),
+		config.ConfigFileUsed(),
```
Can you explain what `ConfigFileUsed` is for? I haven't seen it before

xornivore replied:
This corresponds to this note in the description:
> Note that when secret_backend_command_sha256 is used the datadog.yaml config has to be restricted as follows:
Must be owned by root and cannot have writable others' permissions on *nix.
Cannot be writable by principals other than Local System and Administrators on Windows.

just-chillin commented on pkg/secrets/fetch_secret_test.go:
```diff
@@ -144,6 +150,76 @@ func TestExecCommandError(t *testing.T) {
+		checkConfigRights = prevCheckConfigRights
+	}()
+
+	inputPayload := "{\"version\": \"" + PayloadVersion + "\" , \"secrets\": [\"sec1\", \"sec2\"]}"
```
Nit: might be easier to use `json.Encode` here.

xornivore replied:
This is a bit of copy&paste from earlier tests, but I agree, I think raw strings could help. I can refactor the older tests separately.

just-chillin commented on pkg/secrets/check_rights_windows.go:
```diff
@@ -117,3 +110,150 @@ func checkRights(filename string, allowGroupExec bool) error {
 	return nil
 }
+
+func getACL(filename string) (*winutil.Acl, error) {
```
Mine adding some comments to explain what's going on here - most of the team is linux focused, and I had to spend a bit of time trying to understand what's going on.

xornivore replied:
Yes, for sure.

sgnn7 commented on pkg/secrets/fetch_secret.go:
```diff
@@ -157,3 +191,41 @@ func fetchSecret(secretsHandle []string, origin string) (map[string]string, erro
+	fileSize := stat.Size()
+	if fileSize == 0 {
+		return formatHash(h), nil
+	}
```
Shouldn't this be an error? If the file size is 0, it's probably not a valid binary and we probably shouldn't return a valid hash.

xornivore replied:
Can be done, though It's not an error from the context of the hash calculation.
```
touch /tmp/emptyfile
sha256sum /tmp/emptyfile
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  /tmp/emptyfile
```

clarkb7 replied:
I'm in favor of returning `hash("")` as it is now

sgnn7 commented on pkg/secrets/fetch_secret.go:
```diff
@@ -49,6 +54,35 @@ func execCommand(inputPayload string) ([]byte, error) {
+		}
+
+		if !strings.EqualFold(sha256, secretBackendCommandSHA256) {
+			return nil, fmt.Errorf("error while running '%s': SHA256 mismatch", secretBackendCommand)
```
nit: It would be good here to print both the expected and actual SHAs for easier debugging.

sgnn7 commented on pkg/secrets/check_rights_nix.go:
```diff
@@ -90,3 +91,40 @@ func checkGroupPermission(stat *syscall.Stat_t, usr *user.User, userGroups []str
+var checkConfigRights = func(path string) error {
+	var stat syscall.Stat_t
+	if err := syscall.Stat(path, &stat); err != nil {
+		return fmt.Errorf("invalid config file '%s': can't stat it: %s", path, err)
```
nit: I think `invalid config file` prefix here can be better as the file itself is not "invalid" but its permissions are. Maybe `unable to check permissions on %s` for first one and `invalid config file permissions %s` for the other two?

sgnn7 commented on pkg/secrets/check_rights_nix.go:
```diff
@@ -90,3 +91,40 @@ func checkGroupPermission(stat *syscall.Stat_t, usr *user.User, userGroups []str
 	return nil
 }
+
+var checkConfigRights = func(path string) error {
```
nit: The newly added functions are not in the style of the rest of the file's functions.
nit: We also may want to name this according to it's function (e.g. `ensureRootOwnership` or something similar)

xornivore replied:
> nit: The newly added functions are not in the style of the rest of the file's functions.
Thanks, mind providing some specifics? Missing comments? Name? I am struggling with the name and the `ensureRootOwnership` is not representative of both *nix and Windows cases - maybe `ensureConfigRightsRestricted`? I am following the style of `checkRights` but I need to stub this for tests in a cross platform way. Hence the `= func()`.

sgnn7 replied:
For the style, I would expect them to match the others with `func foo(...) {` rather than `var foo = func(...) {`. If as you mention we're doing it for stubbing, I've seen the [dependency injection pattern](https://github.com/DataDog/datadog-agent/blob/main/pkg/collector/worker/worker.go#L45-L80C27) work pretty well for most cases and it usually results in cleaner code with less bugs since the tests change the minimum needed (though sometimes it requires code reorg).

For the name, you make a good point - maybe `ensureConfigFilePermissions`?

xornivore replied:
I somewhat agree re: dependencies injection, going to defer it to consider refactoring later as it will result in more changes to this already rather complex PR.

I can change to `ensureConfigFilePermissions` or `ensureConfigFilePermissionsRestricted` but these are not in the style of naming in this file. 😄 

There is a precedent of using initialized stubs in this package [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/secrets/secrets.go#L127-L128) and [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/secrets/fetch_secret.go#L112-L113)

sgnn7 replied:
> I somewhat agree re: dependencies injection, going to defer it to consider refactoring later as it will result in more changes to this already rather complex PR. There is a precedent of using initialized stubs in this package [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/secrets/secrets.go#L127-L128) and [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/secrets/fetch_secret.go#L112-L113)

Fair enough though I would say that even if there's precedent, we should try to improve things wherever we can and especially in newly modified code.

> I can change to ensureConfigFilePermissions or ensureConfigFilePermissionsRestricted but these are not in the style of naming in this file. 😄

`checkConfigFilePermissions`?

sgnn7 commented on pkg/secrets/check_rights_nix.go:
```diff
@@ -90,3 +91,40 @@ func checkGroupPermission(stat *syscall.Stat_t, usr *user.User, userGroups []str
+	}
+
+	if stat.Mode&syscall.S_IWOTH != 0 {
+		return fmt.Errorf("invalid config file '%s': others can write", path)
```
nit: `others can write` -> `the file should not have o+w permission`?

sgnn7 commented on pkg/secrets/fetch_secret.go:
```diff
@@ -49,6 +54,35 @@ func execCommand(inputPayload string) ([]byte, error) {
+		}
+
+	} else if hashIsRequired() {
+		return nil, fmt.Errorf("error while running '%s': running elevated and no configured SHA256", secretBackendCommand)
```
```suggestion
		return nil, fmt.Errorf("error while running '%s': running elevated but 'secret_backend_command_sha256' not set", secretBackendCommand)
```

sgnn7 commented on pkg/secrets/check_rights_windows.go:
```diff
@@ -53,34 +47,38 @@ func checkRights(filename string, allowGroupExec bool) error {
+
+	elevated, err := winutil.IsProcessElevated()
+	if err != nil {
+		return fmt.Errorf("unable to determine if running elevated: %s", err)
```
Are we unable to determine if we're elevated or we know that we are not elevated?

xornivore replied:
"unable to determine if we're elevated"

sgnn7 commented on pkg/secrets/check_rights_windows.go:
```diff
@@ -117,3 +115,150 @@ func checkRights(filename string, allowGroupExec bool) error {
+	if _, err := os.Stat(filename); err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("config file %s does not exist", filename)
+		}
```
If `os.Stat` hits errors, shouldn't we unconditionally jump out even if it's not a file-not-exist error?

xornivore replied:
Good catch, this is a copy pasted bit, there is a similar issue in the `checkRights`. Fixing in both places.

sgnn7 commented on pkg/secrets/check_rights_windows.go:
```diff
@@ -117,3 +115,150 @@ func checkRights(filename string, allowGroupExec bool) error {
+	var aclSizeInfo winutil.AclSizeInformation
+	err = winutil.GetAclInformation(fileDacl, &aclSizeInfo, winutil.AclSizeInformationEnum)
+	if err != nil {
+		return fmt.Errorf("could not query ACLs for %s: %s", filename, err)
```
nit
```suggestion
		return fmt.Errorf("could not query ACLs for config file '%s': %s", filename, err)
```

sgnn7 commented on pkg/secrets/check_rights_windows.go:
```diff
@@ -117,3 +115,150 @@ func checkRights(filename string, allowGroupExec bool) error {
+				return fmt.Errorf("invalid permissions for config file '%s': users/groups other than LOCAL_SYSTEM and Administrators have rights on it", filename)
+			}
+		}
+	}
```
This block looks very similar to L85+ lines. Can we extract commonality to a separate func?

xornivore replied:
I thought about it - my concern is that it would obscure reviewing this code. I tried to extract other shared bits like retrieval of SIDs.

sgnn7 reviewed (commented):
@xornivore Left some comments. I'll see if I can get a few more eyes on this from my team given its sensitivity.

cswatt reviewed (approved):
the relnotes look good for docs 👍

xornivore committed:
Address review comments from @just-chillin and @sgnn7 🙌

xornivore committed:
Update pkg/config/config_template.yaml per review comments

hush-hush commented on pkg/secrets/fetch_secret.go:
```diff
@@ -49,6 +54,41 @@ func execCommand(inputPayload string) ([]byte, error) {
 
 	cmd := exec.CommandContext(ctx, secretBackendCommand, secretBackendArguments...)
+
+	if len(secretBackendCommandSHA256) != 0 {
```
```suggestion
	if secretBackendCommandSHA256 != "" {
```

xornivore replied:
This seems like a personal preference here, mind sharing why you feel this is better style?

hush-hush replied:
I think the main logic behind is `len` is more tailored if you plan on iterating over each element of something. While ` != ""` highlight that we're checking for an exact value (empty string here).

It doesn't really matter in the end since the go compiler is going to generate the same code. But I found  ` != ""` clearer since it emphasis that we're working with a string (not a slice or anything else).

it's a detail feel free to ignore.

hush-hush commented on pkg/config/config_template.yaml:
```diff
@@ -701,6 +701,15 @@ api_key:
 
+## @param secret_backend_command_sha256 - string - optional
+## @env DD_SECRET_BACKEND_COMMAND_SHA256 - string - optional
+## `secret_backend_command_sha256` is the SHA256 of the backend command executable.
```
WDYT ?
```suggestion
## `secret_backend_command_sha256` is the expected SHA256 of the backend command executable. The agent will refuse to start if the SHA256 of 'secret_backend_command' is different from the one provided here.
```

xornivore replied:
Sounds good, I will update this description.

hush-hush commented on pkg/config/config_template.yaml:
```diff
@@ -701,6 +701,15 @@ api_key:
+##
+## For more information see: https://github.com/DataDog/datadog-agent/blob/main/docs/agent/secrets.md
+#
+# secret_backend_command_sha256: <SHA256>
```
We might also want to document that the configuration file will be locked as it will prevent upgrading the configuration while the Agent it running.

On that subject, are we OK with that workflow ? Unless I'm misunderstanding the PR, we will no longer be able to update the configuration (with chef, puppet, ansible or other) while the agent is running.
This will for a larger "blind window" for our customer when editing the configuration.

xornivore replied:
We are not locking the configuration file, we are locking the file specified by `secret_backend_command`.

hush-hush replied:
My bad :+1:

hush-hush replied:
One last detail: we should add that we're going to lock the `secret_backend_command` file, since it might prevent customer for updating it while the agent is running.

xornivore commented on pkg/secrets/secrets.go:
```diff
@@ -18,18 +18,25 @@ import (
+	secretBackendTimeout               = defaultSecretBackendTimeout
 	secretBackendCommandAllowGroupExec bool
+	secretBackendCommandSHA256         string
+	configFileUsed                     string
```
This is is the name of the method in the shared config package.

xornivore commented on pkg/secrets/secrets.go:
```diff
@@ -40,12 +47,14 @@ func init() {
 	secretBackendTimeout = timeout
 	SecretBackendOutputMaxSize = maxSize
 	secretBackendCommandAllowGroupExec = groupExecPerm
+	secretBackendCommandSHA256 = sha256
+	configFileUsed = configFile
```
Happy to do it in a separate PR.

hush-hush replied:
:+1: for another PR, but we want to target `7.42.0` too so we can troubleshoot this feature is needed.

xornivore commented on releasenotes/notes/secret_backend_command_sha256-b322140f976f02e1.yaml:
```diff
@@ -0,0 +1,8 @@
+
+enhancements:
+  - |
+    Allows running secrets in the process-agent on Windows as long as the user provides secret_backend_command_sha256 and restricts access to the datadog.yaml config file.
```
This addition belongs in the section above, as here I am providing ability to use secrets for process agent on Windows (unable to use them there today).

hush-hush replied:
```suggestion
    Allows running secrets in the process-agent on Windows as long as the user provides ``secret_backend_command_sha256`` and restricts access to the datadog.yaml config file.
```

xornivore commented on pkg/secrets/check_rights_nix.go:
```diff
@@ -90,3 +92,42 @@ func checkGroupPermission(stat *syscall.Stat_t, usr *user.User, userGroups []str
+		return nil, err
+	}
+
+	if err = syscall.Flock(fd, syscall.LOCK_EX); err != nil {
+		syscall.Close(fd)
+		return nil, err
```
Yes https://linux.die.net/man/2/flock
> Furthermore, the lock is released either by an explicit LOCK_UN operation on any of these duplicate descriptors, or when all such descriptors have been closed.

xornivore commented on pkg/secrets/check_rights_nix.go:
```diff
@@ -90,3 +92,42 @@ func checkGroupPermission(stat *syscall.Stat_t, usr *user.User, userGroups []str
+	return false, nil
+}
+
+// lockOpenFile opens the file and prevents overwrite and delete by another process
+var lockOpenFile = func(path string) (*os.File, error) {
```
I will change that, I don't need to stub this.

xornivore commented on pkg/secrets/check_rights_windows.go:
```diff
@@ -117,3 +116,155 @@ func checkRights(filename string, allowGroupExec bool) error {
+var hashIsRequired = winutil.IsProcessElevated
+
+// lockOpenFile opens the file and prevents overwrite and delete by another process
+func lockOpenFile(name string) (*os.File, error) {
+	h, err := winOpenFileShareRead(name)
+	if err != nil {
+		return nil, err
+	}
+	return os.NewFile(uintptr(h), name), nil
+}
```
Yes, the handle will be released.

xornivore committed:
Address review comments from @hush-hush and @sgnn7

hush-hush commented on releasenotes/notes/secret_backend_command_sha256-b322140f976f02e1.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+features:
+  - |
+    Adds support for `secret_backend_command_sha256` SHA for the `secret_backend_command` executable. If `secret_backend_command_sha256` is used,
+    the following restrictions are in place:
+    - Value specified in the `secret_backend_command` setting must be an absolute path.
+    - Permissions for the datadog.yaml config file must disallow write access by unprivileged accounts.
+    The agent will refuse to start if the actual SHA256 of 'secret_backend_command' executable is different from the one specified by `secret_backend_command_sha256`.
```
We need double backticks since `reno` produce a RST in the end for the CHANGELOG.
```suggestion
    Adds support for ``secret_backend_command_sha256`` SHA for the ``secret_backend_command`` executable. If ``secret_backend_command_sha256`` is used,
    the following restrictions are in place:
    - Value specified in the ``secret_backend_command`` setting must be an absolute path.
    - Permissions for the datadog.yaml config file must disallow write access by unprivileged accounts.
    The agent will refuse to start if the actual SHA256 of ``secret_backend_command`` executable is different from the one specified by ``secret_backend_command_sha256``.
```

xornivore replied:
Ah, thank you so much!

hush-hush reviewed (approved):
Added a few nitpicks but looks good overall.

xornivore committed:
Address review comments from @hush-hush

Zenithar commented on pkg/secrets/fetch_secret.go:
```diff
@@ -157,3 +197,41 @@ func fetchSecret(secretsHandle []string, origin string) (map[string]string, erro
+	buffer := make([]byte, bufSize)
+
+	for {
+		read, err := f.Read(buffer)
```
It would be best if you implemented a limiter to ensure that the agent can't be blocked at the bootstrap time while it tries to hash the target file.

Also, consider migrating to `io.CopyN` to read and stream directly.

xornivore replied:
There is a timeout in place for execution of secret backends [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/secrets/fetch_secret.go#L47-L48). It will include the hashing part. Default is 5s. My fear of using `CopyN` is in making it more difficult to disambiguate the source of errors.

Zenithar replied:
I was looking more for a memory limiter than a time limiter. Currently, if the user set `/dev/zero` as the target script, the sha256 computation will block due to the infinite nature of `/dev/zero` (aside from its type check). This could be used to prevent an agent from starting, and in the worst case, it could generate a heavy workload on the computer itself. You could add an acceptable script file size check above and ensure the reader is truncated/ended by an `io.LimitReader`.

Zenithar commented on pkg/secrets/fetch_secret_test.go:
```diff
@@ -144,6 +152,76 @@ func TestExecCommandError(t *testing.T) {
 
+func TestExecCommandWithHash(t *testing.T) {
+	prevCheckConfigFilePermissions := checkConfigFilePermissions
+	checkConfigFilePermissions = func(string) error { return nil }
```
These operations could generate flaky test results if the tests are executed in parallel. You should migrate to local variables, and stop monkey-patching global instances.

xornivore replied:
Thank you @Zenithar. Point taken, this was brought up in other comments, my plan is to change these, but in a separate PR.

Zenithar commented on pkg/secrets/fetch_secret.go:
```diff
@@ -49,6 +54,41 @@ func execCommand(inputPayload string) ([]byte, error) {
+	if secretBackendCommandSHA256 != "" {
+		if !filepath.IsAbs(secretBackendCommand) {
+			return nil, fmt.Errorf("error while running '%s': absolute path required with SHA256", secretBackendCommand)
+		}
```
Please consider checking if the target file is a symlink too. The agent must not be able to run a target binary pointing to another one via a symlink. Symbolic link attacks can be used to change a file's permissions, corrupt data, destroy data, or as a DoS attack.

xornivore replied:
Great point, will add this check, this however will only take place when `secret_backend_command_sha256` is provided.

xornivore replied:
Added this check.

xornivore committed:
Separate running secrets in process-agent from sha256 feature

- Put back removal of `secrets` package from build in `process-agent`.
- Remove "hash required" check - optional feature at this point.
- Allow `ddagentuser` writeable `datadog.yaml` config on Windows.
- Remove unneeded check for elevation and hash.
- Update release note.

xornivore committed:
Address feedback from @Zenithar

- Add io.LimitReader and check for hashable content max size.
- Add check for file mode to prevent non-regular file types.

clarkb7 reviewed (approved):
I think no longer necessary because of https://github.com/DataDog/datadog-agent/pull/14628

but LGTM for Windows

Zenithar commented on pkg/secrets/fetch_secret.go:
```diff
@@ -157,3 +196,52 @@ func fetchSecret(secretsHandle []string, origin string) (map[string]string, erro
+	}
+
+	if !stat.Mode().IsRegular() ||
+		(stat.Mode()&(os.ModeSymlink|os.ModeSocket|os.ModeCharDevice|os.ModeDevice|os.ModeNamedPipe) != 0) {
```
👍

Zenithar reviewed (approved):
LGTM

taraxvoid reviewed (approved):
LGTM

sgnn7 commented:
+1 that it may no longer be necessary side because of https://github.com/DataDog/datadog-agent/pull/14628. If we want the sha256 validation by itself, we should probably move the elevation/locking checks to a separate PR.

xornivore commented:
> +1 that it may no longer be necessary side because of #14628. If we want the sha256 validation by itself, we should probably move the elevation/locking checks to a separate PR.

The logic to handle "sandboxing" to "ddagentuser" was moved to https://github.com/DataDog/datadog-agent/pull/14628. I will rebase this PR once the other is in `main`.

xornivore committed:
Fix rebase

just-chillin reviewed (approved):
:shipit:

just-chillin reviewed (approved):
:shipit:

xornivore committed:
Add sha256 output to the secrets debug info for CLI
